src/python/pants/core/goals/generate_lockfiles.py

```diff
@@ -407,6 +430,25 @@ async def generate_lockfiles_goal(
     return GenerateLockfilesGoal(exit_code=0)
 
 
+def _preferred_environment(request: GenerateLockfile, default: EnvironmentName) -> EnvironmentName:
+
+    if not isinstance(request, GenerateLockfileWithEnvironments):
```

Comment:
An alternative to subtyping might be to give `GenerateLockfile` a field which is optional (for now?)
```python
  environments: tuple[EnvironmentName, ...] |  None = None
```

Reply:
Alas, it's not possible to have an optional value there, because `GenerateLockfile` has subclasses with required parameters. `dataclasses` doesn't make this case easy.